src/net/mod.rs

```diff
@@ -6,21 +6,145 @@
+    T: Serialize,
+{
+    /// Construct Stabilizer's default network users.
+    ///
+    /// # Args
+    /// * `stack` - The network stack that will be used to share with all network users.
+    /// * `phy` - The ethernet PHY connecting the network.
+    /// * `cycle_counter` - The clock used for measuring time in the network.
+    /// * `app` - The name of the application.
+    /// * `mac` - The MAC address of the network.
+    ///
+    /// # Returns
+    /// A new struct of network users.
+    pub fn new(
+        stack: NetworkStack,
+        phy: EthernetPhy,
+        cycle_counter: CycleCounter,
+        app: &str,
+        mac: smoltcp_nal::smoltcp::wire::EthernetAddress,
+    ) -> Self {
+        let stack_manager =
+            cortex_m::singleton!(: NetworkManager = NetworkManager::new(stack))
+                .unwrap();
+
+        let processor = NetworkProcessor::new(
+            stack_manager.acquire_stack(),
+            phy,
+            cycle_counter,
+        );
+
+        let prefix = get_device_prefix(app, mac);
+
+        let settings = MiniconfClient::new(
+            stack_manager.acquire_stack(),
+            &get_client_id(app, "settings", mac),
+            &prefix,
+        );
+
+        let telemetry = TelemetryClient::new(
+            stack_manager.acquire_stack(),
+            &get_client_id(app, "tlm", mac),
+            &prefix,
+        );
+
+        NetworkUsers {
+            miniconf: settings,
+            processor,
+            telemetry,
+        }
+    }
+
+    /// Update and process all of the network users state.
+    ///
+    /// # Returns
+    /// An indication if any of the network users indicated a state change.
+    pub fn update(&mut self) -> UpdateState {
+        // Poll for incoming data.
+        let poll_result = self.processor.update();
+
+        // Update the MQTT clients.
+        self.telemetry.update();
+
+        match self.miniconf.update() {
+            UpdateState::Updated => UpdateState::Updated,
+            UpdateState::NoChange => poll_result,
+        }
+    }
+}
+
+/// Get an MQTT client ID for a client.
+///
+/// # Args
+/// * `app` - The name of the application
+/// * `client` - The unique tag of the client
+/// * `mac` - The MAC address of the device.
+///
+/// # Returns
+/// A client ID that may be used for MQTT client identification.
+fn get_client_id(
+    app: &str,
+    client: &str,
+    mac: smoltcp_nal::smoltcp::wire::EthernetAddress,
+) -> String<consts::U64> {
+    let mac_string = {
+        let mut mac_string: String<consts::U32> = String::new();
+        let mac = mac.as_bytes();
+
+        // Note(unwrap): 32-bytes is guaranteed to be valid for any mac address, as the address has
+        // a fixed length.
+        write!(
+            &mut mac_string,
+            "{:02x}-{:02x}-{:02x}-{:02x}-{:02x}-{:02x}",
+            mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]
+        )
+        .unwrap();
+
+        mac_string
+    };
```

Comment:
`EthernetAddress` already implements `Display`, does that not format it well?

Reply:
Actually, the code is basically the same: https://docs.rs/smoltcp/0.4.0/src/smoltcp/wire/ethernet.rs.html#70-76

Reply:
I didn't know it implemented display - definitely going to use that. Thanks for pointing this out